src/web/utils/parserUtils.ts

```diff
@@ -0,0 +1,303 @@
+import BrowserUtils from './browserUtils';
+import type {MarkdownTextInputElement} from '../../MarkdownTextInput.web';
+import {addNodeToTree, buildTree} from './treeUtils';
+import type {NodeType, TreeNode} from './treeUtils';
+import type {PartialMarkdownStyle} from '../../styleUtils';
+import {getCurrentCursorPosition, moveCursorToEnd, setCursorPosition} from './cursorUtils';
+import {addStyleToBlock} from './blockUtils';
+
+type MarkdownType = 'bold' | 'italic' | 'strikethrough' | 'emoji' | 'link' | 'code' | 'pre' | 'blockquote' | 'h1' | 'syntax' | 'mention-here' | 'mention-user' | 'mention-report';
+
+type MarkdownRange = {
+  type: MarkdownType;
+  start: number;
+  length: number;
+  depth?: number;
+};
+
+type Paragraph = {
+  text: string;
+  start: number;
+  length: number;
+  markdownRanges: MarkdownRange[];
+};
+
+function ungroupRanges(ranges: MarkdownRange[]): MarkdownRange[] {
+  const ungroupedRanges: MarkdownRange[] = [];
+  ranges.forEach((range) => {
+    if (!range.depth) {
+      ungroupedRanges.push(range);
+    }
+    const {depth, ...rangeWithoutDepth} = range;
+    Array.from({length: depth!}).forEach(() => {
+      ungroupedRanges.push(rangeWithoutDepth);
+    });
+  });
+  return ungroupedRanges;
+}
+
+function splitTextIntoLines(text: string): Paragraph[] {
+  let lineStartIndex = 0;
+  const lines: Paragraph[] = text.split('\n').map((line) => {
+    const lineObject: Paragraph = {
+      text: line,
+      start: lineStartIndex,
+      length: line.length,
+      markdownRanges: [],
+    };
+    lineStartIndex += line.length + 1; // Adding 1 for the newline character
+    return lineObject;
+  });
+
+  return lines;
+}
+
+function mergeLinesWithMultilineTags(lines: Paragraph[]) {
+  let multiLineRange: MarkdownRange | null = null;
+  let lineWithMultilineTag: Paragraph | null = null;
+  let i = 0;
+  while (i < lines.length) {
```

Comment:
I used `while` because I want to increase the number in the `i` variable only when the condition is false. For me while the loop is more clear in such cases